lib/trino-filesystem-s3/src/main/java/io/trino/filesystem/s3/S3FileSystemStats.java

```diff
@@ -0,0 +1,263 @@
+    private final AwsSdkV2ApiCallStats abortMultipartUpload = new AwsSdkV2ApiCallStats();
+    private final AwsSdkV2ApiCallStats uploadPart = new AwsSdkV2ApiCallStats();
+
+    private static final AwsSdkV2ApiCallStats dummy = new DummyAwsSdkV2ApiCallStats();
+
+    @Managed
+    @Nested
+    public AwsSdkV2ApiCallStats getTotal()
+    {
+        return total;
+    }
+
+    @Managed
+    @Nested
+    public AwsSdkV2ApiCallStats getHeadObject()
+    {
+        return headObject;
+    }
+
+    @Managed
+    @Nested
+    public AwsSdkV2ApiCallStats getGetObject()
+    {
+        return getObject;
+    }
+
+    @Managed
+    @Nested
+    public AwsSdkV2ApiCallStats getListObjectsV2()
+    {
+        return listObjectsV2;
+    }
+
+    @Managed
+    @Nested
+    public AwsSdkV2ApiCallStats getPutObject()
+    {
+        return putObject;
+    }
+
+    @Managed
+    @Nested
+    public AwsSdkV2ApiCallStats getDeleteObject()
+    {
+        return deleteObject;
+    }
+
+    @Managed
+    @Nested
+    public AwsSdkV2ApiCallStats getDeleteObjects()
+    {
+        return deleteObjects;
+    }
+
+    @Managed
+    @Nested
+    public AwsSdkV2ApiCallStats getCreateMultipartUpload()
+    {
+        return createMultipartUpload;
+    }
+
+    @Managed
+    @Nested
+    public AwsSdkV2ApiCallStats getCompleteMultipartUpload()
+    {
+        return completeMultipartUpload;
+    }
+
+    @Managed
+    @Nested
+    public AwsSdkV2ApiCallStats getAbortMultipartUpload()
+    {
+        return abortMultipartUpload;
+    }
+
+    @Managed
+    @Nested
+    public AwsSdkV2ApiCallStats uploadPart()
+    {
+        return uploadPart;
+    }
+
+    public MetricPublisher newMetricPublisher()
+    {
+        return new JmxMetricPublisher(this);
+    }
+
+    public static final class JmxMetricPublisher
+            implements MetricPublisher
+    {
+        private static final Set<SdkMetric<?>> ALLOWED_METRICS = Set.of(API_CALL_SUCCESSFUL, RETRY_COUNT, API_CALL_DURATION, ERROR_TYPE);
+
+        private static final Logger log = Logger.get(JmxMetricPublisher.class);
+
+        private final S3FileSystemStats stats;
+
+        public JmxMetricPublisher(S3FileSystemStats stats)
+        {
+            this.stats = requireNonNull(stats, "stats is null");
+        }
+
+        @Override
+        public void publish(MetricCollection metricCollection)
+        {
+            try {
+                Optional<String> serviceId = metricCollection.metricValues(SERVICE_ID).stream().filter(Objects::nonNull).findFirst();
+                Optional<String> operationName = metricCollection.metricValues(OPERATION_NAME).stream().filter(Objects::nonNull).findFirst();
+                if (serviceId.isEmpty() || operationName.isEmpty()) {
+                    log.warn("ServiceId or OperationName is empty for AWS MetricCollection: %s", metricCollection);
+                    return;
+                }
+
+                if (!serviceId.get().equals("S3")) {
```

Comment:
Thank you for the review. As far as I tested for multiple operations (e.g. headObject, getObject) it was always `S3` in uppercase. And since it represents AWS service name, I think there's no other value for this.